### ISSUE TITLE
Chromium-based apps: 'delete surrounding text' doesn't work

### ISSUE THREAD

**Issue by rik-shaw** [ibus/ibus#2354]
Which distribution and version?: 
Ubuntu 20.04

Which desktop environment and version?: 
GNOME Shell 3.36.9, xfce4-session 4.16.0 (Xfce 4.16), Cinnamon 4.8.6

Which session type?: 
X11

Which application and version?: 
Chromium and Chromium-based browsers like Brave, plus Electron apps (Skypeforlinux, signal, telegram, mailspring, etc)

IBus version?: 
IBus 1.5.22

Issue description:
It seems that "delete surrounding text" is not working correctly: For example, we use `;n` to produce `ŋ` but in Chromium-based apps (browsers and electron apps) the `;` is not deleted, meaning that `;ŋ` is still displayed. We use the `ibus-kmfl` engine but I think the same issue is there for other engines also?

Steps to reproduce:
1. Type any sequence that would require a character to be deleted and replaced with another character
1. The first character is not deleted 

Can you reproduce your problem when you restart ibus-daemon? (yes / no): 
yes

Do you see any errors when you run ibus-daemon with the verbose option?:
no

Can you reproduce your problem with a new user account instead of the current your account? (yes / no): 
yes

NOTE: our kmfl packages are in the `packages.sil.org` repository, so please let me know if you can't reproduce this issue with engines you are more familiar with, at this point I am not sure if it is `ibus` or `ibus-kmfl` where the issue is. Thank you again for your great work over the years!

**Comment by mike-fabian** [comment]
This very simple use case of surrounding text, i.e. delete a single string somewhere to the left of the cursor, is working for me in gedit but **not** in google-chrome.

Versions:

```
[mfabian@fedora ~]$ cat /etc/fedora-release 
Fedora release 34 (Thirty Four)
[mfabian@fedora ~]$ rpm -q ibus
ibus-1.5.24-5.fc34.x86_64
[mfabian@fedora ~]$ rpm -q xfce4-session
xfce4-session-4.16.0-3.fc34.x86_64
[mfabian@fedora ~]$ rpm -q ibus-typing-booster
ibus-typing-booster-2.14.12-1.fc34.noarch
[mfabian@fedora ~]$ rpm -q gedit
gedit-40.1-1.fc34.x86_64
[mfabian@fedora ~]$ rpm -q google-chrome-unstable
google-chrome-unstable-95.0.4638.17-1.x86_64
[mfabian@fedora ~]$ 
```

Video:


https://user-images.githubusercontent.com/2330175/134387316-233e6e1e-97b4-445c-ad47-5f923162606c.mp4

**Comment by rik-shaw** [comment]
@mike-fabian thank you very much for your quick confirmation. Yes, for me as well, gedit or basically *all* non-Chrom(ium) based apps work as expected, but with many Electron apps now getting to be more common it is getting to be a growing issue for many of our partners that keyboard Ethiopic scripts and latin-based minority scripts that rely on the "delete surrounding text" feature for being able to trigger different base characters. The example I use of `;n` may seem odd to many, but for our users it is helpful to have the `;` display (and then be deleted after hitting `n`) so that they aren't confused about if they hit the first key or not (many are new to computers and new to typing).

**Comment by mike-fabian** [comment]
If it is only that you want to keep displaying the `;` while the sequence is not yet finished, I think I would just not commit the `;` but keep it in preedit.  Like for example the “latn-pre (m17n)” input method of ibus-m17n does (ibus-typing-booster does this as well when using latn-pre). Looks like this and works both in gedit and google-chrome:


https://user-images.githubusercontent.com/2330175/134422147-cad17eca-6d96-425b-a072-726735c48c6f.mp4

**Comment by mike-fabian** [comment]
You might consider that for `;n` to produce `ŋ`, keep the `;` in preedit until the `n` is typed. Then you don‘t need to delete it from the application using surrounding text, it is enough to remove the `;` from preedit then. The preedit is easier to control.

Surrounding text is unfortunately still very buggy and the bugs differ depending on the toolkit.

I hope the bugs in surrounding text can be fixed because there are some things  where surrounding text is absolutely necessary. 

For just converting `;n` to `ŋ` it is not really needed. But if you want something like this:

There is already an `;` somewhere in the application and then you position the cursor behind that `;` and type `n`. If you want that this becomes an `ŋ` even in that situation where the cursor has been moved behind the `;`, then you need the surrounding text.

To improve reliability and make it work better in more applications, I would try to avoid surrounding text when you can do it by keeping the preedit and use surrounding text only when there is no other way.

But of course it is important to fix surrounding text, this really should work better.

**Comment by rik-shaw** [comment]
@mike-fabian thanks again for your quick and helpful reply. The framework we use for writing keyboards is `[Keyman](https://github.com/keymanapp/keyman)` so I am not immediately certain if we can take advantage of preedit or not. But I do understand that "delete surrounding text" is complicated, in fact if I can move away from it then things like wine applications, etc become more possible.

From my view, it is acceptable to NOT have a pre-existing `;` have an `n` later placed after it. The user will need to remove the `;` and again type `;n` to produce `ŋ`.

@mcdurdin or @darcywong00 can you comment on if this is possible in Keyman to use "preedit" for our visible triggers rather than needing "delete surrounding text"? It would likely help with all sorts of headaches in Linux keyboarding if we don't need to rely on "delete surrounding text"...

**Comment by rik-shaw** [comment]
@mike-fabian sorry I should mention also that many keyboards such as ones for Ethiopic scripts need to create digraphs (single characters representing a consonant - vowel combination). Here is how we type in it, for example:

`m` produces the "base" which is often used "as is" (it is the most common digraph): `ም`
Then pressing a vowel, such as this `m` `a` will result in `ማ`. So the `ም` appears but then is 'deleted' and replaced with `ማ` after the `a` is pressed. Do you think that the preedit approach you describe could handle this scenario?

**Comment by mcdurdin** [comment]
This would be algorithmically complex to achieve with Keyman, as the baseline assumption is that surrounding text is always editable. For trivial keyboards this may be viable, but many keyboards rely on building up multiple characters over time and there is never a clear breakpoint to commit preceding text. This is particularly true for some Indic and many South-East Asian scripts.

We examined the possiblity of working with a pretext or "compose buffer" on Windows years ago, using Windows' IME composition model, and came to the conclusion that it would not be generalisable, and so instead put effort into improving support for editing surrounding (before-only) text.

**Comment by mcdurdin** [comment]
Is there a Chromium bug report for this? I've found the Chromium team to be quite responsive to addressing these kinds of bugs.

**Comment by mcdurdin** [comment]
Tagging @ermshiperete.

**Comment by mike-fabian** [comment]
> @mike-fabian sorry I should mention also that many keyboards such as ones for Ethiopic scripts need to create digraphs (single characters representing a consonant - vowel combination). Here is how we type in it, for example:
> 
> `m` produces the "base" which is often used "as is" (it is the most common digraph): `ም`
> Then pressing a vowel, such as this `m` `a` will result in `ማ`. So the `ም` appears but then is 'deleted' and replaced with `ማ` after the `a` is pressed. Do you think that the preedit approach you describe could handle this scenario?

Yes. This video shows it using latn-post and am-sera as examples:

https://user-images.githubusercontent.com/2330175/134472020-260eb663-8055-4a25-b0cd-9727e13c9e9c.mp4

**Comment by mike-fabian** [comment]
> This would be algorithmically complex to achieve with Keyman, as the baseline assumption is that surrounding text is always editable. For trivial keyboards this may be viable, but many keyboards rely on building up multiple characters over time and there is never a clear breakpoint to commit preceding text. This is particularly true for some Indic and many South-East Asian scripts.

I think one can do both, it doesn’t have to be 100% one way or the other.

Most languages uses spaces between words (except for Japanese and Chinese) so the end of a word is a clear breakpoint where one can certainly commit.

That’s the way ibus-typing-booster does it, it keeps the preedit until a word is finished (it needs to to that because it wants to show possible completions for a partly typed word).
ibus-m17n commits earlier like you see in my video using ibus-m17n with am-sera, typing `m` keeps the preedit, typing the `a` after that commits. Also for Indian languages like Hindi, ibus-m17n commits as soon as some cluster of characters which may influence each other is finished.

One can do both things of course, i.e. when typing that `m`, don’t commit, keep the preedit, wait what comes next, if an `a` is typed combine it with the `m`.

And if there is already an ም somewhere in the application and one moves the cursor behind that and types `a`, only then try to use surrounding text to delete the ም and replace it with ማ.

Using both approaches, i.e. keep the preedit if possible  and use surrounding text if keeping the preedit was not possible would make it work already in many cases where  it doesn’t work now.

I think the surrounding text approach is mostly needed to fix things up when moving the cursor around in already existing text. Which is of course also an important feature, so of course surrounding text needs to be fixed and improved.

But keeping the preedit until some breakpoint is reached is also a good thing.

> We examined the possiblity of working with a pretext or "compose buffer" on Windows years ago, using Windows' IME composition model, and came to the conclusion that it would not be generalisable, and so instead put effort into improving support for editing surrounding (before-only) text.

I think doing both is better, work with the preedit where it makes sense **and** use surrounding text when not.

**Comment by rik-shaw** [comment]
Am I correct that this Chromium issue is the relevant bug? https://bugs.chromium.org/p/chromium/issues/detail?id=1123174

If so, it appears that it is a bit dormant as they don't have feedback on any input methods that need this feature. So, I could attempt to add to that issue, but as developers would @mike-fabian or @mcdurdin be able to more effectively contribute there? Certainly I (and the users I support) are the most motivated to resolve this issue, but wondering if the technical understanding from you would be more efficient in communicating to the Chromium devs?

**Comment by mike-fabian** [comment]
I added two comments there as wel.

@rik-shaw By the way, does get_surrounding_text() work for you in google-chrome? I just tested with ibus-typing-booster and get_surrounding_text() always returned
the last valid surrounding text from the previous window which had focus for me.

For example, if I have the text ”This is a test|” in firefox where “|” indicates the cursor position, then switch focus to google-chrome and type something there and look at my debug output, I see that I still get the old surrounding text “This is a test|” from the firefox window which previously had focus.

**Comment by mcdurdin** [comment]
> Most languages uses spaces between words (except for Japanese and Chinese) so the end of a word is a clear breakpoint where one can certainly commit.

Many South-East Asian languages do not use spaces between words either. Many polysynthetic indigenous languages have very long words where preedit-style input feels awkward (as far as I understand, the app does not process preedit text at all until it is committed, so things like search boxes with dynamic search would lose functionality, etc).

But all that aside, Keyman keyboards do not have clear breakpoints that can be easily determined, and as there are thousands of Keyman keyboard layouts available, it's not something we can engineer into them now. We haven't had to implement a preedit or composition design on any other platform so I am really keen to try and resolve the platform or app integration issues rather than back ourselves into a corner with a half-way solution using preedit.

@ermshiperete leads the Linux integration but is on leave currently, back next week; he'll be able to engage further.

**Comment by mike-fabian** [comment]
> Many South-East Asian languages do not use spaces between words either. Many polysynthetic indigenous languages have very long words where preedit-style input feels awkward (as far as I understand, the app does not process preedit text at all until it is committed, so things like search boxes with dynamic search would lose functionality, etc).

That depends on the app. Some apps do process preedit text before it is comitted, for example chrome does. In this screenshot you see the letter “y” typed with ibus-m17n latn-post still in preedit (note the underline) because it is waiting for the next key to possibly combine it with the y. Even though the “y”  is in preedit, google-chrome processes it for the search, you see search results starting with “y”:

![Screenshot](https://user-images.githubusercontent.com/2330175/134629786-b1f42658-e24a-4bcb-9158-e37f5e6c2e19.png)

But that really depends on the app, firefox does **not** do that for example.

On the other hand, firefox implements delete-surrounding-text whereas chrome doesn't.

**Comment by mcdurdin** [comment]
See #2436 for a related fix.

**Comment by rik-shaw** [comment]
@mike-fabian have you had a chance to see if this above mentioned PR (#2436) is a fix for your use case? Also note there are a few additional PRs from @ermshiperete that may work together to correct the issues: #2438 and #2440

**Comment by ermshiperete** [comment]
Yes, Chrome doesn't support surrounding text yet. However, ibus reports the IBUS_CAP_SURROUNDING_TEXT capability for chromium-based apps. #2436 improves the ibus part so that Chrome no longer gets reported as supporting surrounding text.

**Comment by fujiwarat** [comment]
@ermshiperete I don't know what you try to fix with that patch but reporting the warning is good but specifying "firefox" is very bad.

**Comment by mcdurdin** [comment]
@fujiwarat, I don't understand your attitude toward the patch in #2436: it seems unnecessarily antagonistic.

The original patch that you wrote in 7b3b8c8 made a specific fix just for firefox, which broke things for many, many other applications. You made it so that *all* applications reported `IBUS_CAP_SURROUNDING_TEXT`, even if they did not support it.

That's very bad.

You either need to wind that back or else find some way to target that patch to work just for firefox.

In my opinion, the patch in #2436 is a good, simple patch. The patch is specific, targeted, and is trivial to to wind back later if/when Firefox resolve the inconsistent behavior that caused you to write the patch in the first place. Can you explain why you think this patch is "very bad"?

**Comment by fujiwarat** [comment]
> The original patch that you wrote in a specific fix just for firefox,

No, this patch works for chorme too.

Can you explain why you think this new patch in is "very bad"?

1. Hard-coded "firefox" is very bad and this patch is not for firefox only.
2. Changing caps after _request_surrounding_text() is very bad
3. If your application does not support surrounding text, fixing it is the right way and emitting "retrieve-surrounding-text" siginal is no problem for those appliations.

**Comment by mcdurdin** [comment]
> No, this patch works for chorme too.

No, it doesn't. This patch has completely broken the behaviour for Chrome, which is the _entire_ point of this issue. This has been explained in detail in #2436. 

> * Hard-coded "firefox" is very bad and this patch is not for firefox only.

You still haven't explained why a hardcoded "firefox" is very bad. Personally, I don't like hardcoded behaviours to work around app problems, but it's a pretty standard technique for dealing with compatibility. We use it when we have to, and when (if!) the apps improve, we back it out again.

Which other apps have improved behaviour with this patch? There's no documentation anywhere about any other apps in the patch -- it specifically cites firefox, and nothing else.

> * Changing caps after _request_surrounding_text() is very bad

That was the original behaviour in versions 1.5.19 and earlier:

https://github.com/ibus/ibus/blob/a40631e166137c9042a68c2d76844e7afc53d388/client/gtk2/ibusimcontext.c#L386-L392

Your patch removed that capability update, and landed in 1.5.20. So, if it was okay to do this before 1.5.20, then why is it not okay now?

> * If your application does not support surrounding text, fixing it is the right way and emitting "retrieve-surrounding-text" siginal is no problem for those appliations.

By that argument, you should be patching firefox, not patching ibus to work around a firefox bug.

Yes, Chrome should implement surrounding text support, but there are thousands of apps without surrounding text support -- are you planning to patch them all? The way it stands, ibus is broken for all those apps, for many, many use cases, as documented in this very issue.

Here's another way forward: we can open a PR to revert the original patch 7b3b8c8b0c6a41ab524e0be9474825da9cba96ac, and then put effort into getting the firefox bug properly resolved. How does that sound?

**Comment by fujiwarat** [comment]
>  but it's a pretty standard technique for dealing with compatibility. 

That will be consider by me. I'm asking which problem you try to fix with detail at first in #2436 but it refers this issue again.

What is the IPA (SIL) keyboard? The upstream seems not to relative wit ibus. What is ibus-kmfl?

**Comment by mcdurdin** [comment]
Thank you for your reply @fujiwarat, I appreciate your engagement with us on this issue.

Summarising the bug: the patch 7b3b8c8 causes ibus to _always_ report `IBUS_CAP_SURROUNDING_TEXT` for _every_ application, even when the application doesn't support surrounding-text.

> What is the IPA (SIL) keyboard?

The [IPA (SIL) keyboard](https://github.com/keymanapp/keyboards/tree/master/release/sil/sil_ipa) is a keyboard layout that runs on the open source multi-platform [Keyman](https://github.com/keymanapp/keyman) input method framework. Keyman has a library of keyboard layouts supporting over 2,000 languages.

On Linux, Keyman is implemented as an ibus input method. Keyman's input model requires either surrounding-text support, or the ability to emit backspaces to remove existing characters from the text store for text transformation at input time. (There's more discussion on the reasons for this earlier in the issue history.)

> What is ibus-kmfl?

ibus-kmfl is the legacy version of Keyman for Linux (kmfl == Keyman for Linux). The new name is 'ibus-keyman', but the ibus integration code is largely the same.

> I'm asking which problem you try to fix

The problem that #2436 is trying to fix is that the Keyman input method needs to be able to detect if the target app supports surrounding-text. If it does, then Keyman can use the surrounding-text APIs to transform the content in the text store. If the app doesn't support surrounding-text, then Keyman needs to use a much less elegant approach of injecting backspaces into the key event stream, and then emitting the transformed content through the regular APIs.

We want to be able to use surrounding-text if it is available. It's a lot cleaner, and has fewer potential breakage points -- for example, the app can [choose to delete more than one character](https://github.com/keymanapp/keyman/issues/7130#issuecomment-1228602252) with backspace, and Keyman can't know it did that. But the reality is that there are too many apps that don't support surrounding-text today, and will be for a very long time. So we need to be able to fall back to the less elegant approach, even if it doesn't cover 100% of all possible cases. Supporting more apps is always better than supporting fewer apps.

Fallback is not possible if we cannot detect `IBUS_CAP_SURROUNDING_TEXT` for the target app. The patch 7b3b8c8 causes ibus to _always_ report `IBUS_CAP_SURROUNDING_TEXT` for _every_ application, even when the application doesn't support surrounding-text. Thus, in #2436, we decided to try and get the best of both worlds: continue to work around the current issue in firefox that was the motivation for 7b3b8c8, without breaking the detection of surrounding-text for all other apps, including Chromium-based apps.

**Comment by fujiwarat** [comment]
Thank you for your explanation of your issue.
I see you use https://github.com/keymanapp/keyman/tree/master/linux/ibus-keyman and I can consider ibus-kmfl as ibus-keyman .

Reading your problem, seems you wish to update the capabilities if emitting "retrieve-surrounding" signal is failed so that engines disables the surrounding text feature but I deleted the logic since the second typing could be succeeded.

So if the updating capabilities is useful for engines and checking the second typing is a specific case of firefox, I think the firefox issue should be fixed in firefox.
Probably you don't agree with updating capabilities after the second typing but wish to update the capabilities with the first typing immediately and then I think simply reverting the original patch would be better. That's why indicating "firefox" in the code is bad.

Now I tried firefox and I cannot reproduce the original issue in my test repo.

**Comment by mcdurdin** [comment]
@fujiwarat Thank you 😄! I am away until Monday now but look forward to bringing this to a good resolution.

**Comment by fujiwarat** [comment]
I created the tentative patch. Could you check the branch?

**Comment by ermshiperete** [comment]
Thanks @fujiwarat ! The changes in your branch look good. I tested them and I get the same results as with #2436. :+1:

**Comment by mcdurdin** [comment]
Change looks good @fujiwarat. Thank you!

@ermshiperete can we contribute by trying to bisect and find the firefox build (and thence the patch) which fixed the behaviour which necessitated the original patch? It'd be good to be able to document why the patch is no longer needed.

**Comment by rik-shaw** [comment]
Thanks again for everyone's involvement in sorting this issue out, it will be great to have ibus keyboarding working in Chromium based apps. I really appreciate all of your extra efforts.

**Comment by ermshiperete** [comment]
# Bisecting Firefox for ibus bug #2054

https://github.com/ibus/ibus/issues/2054

Download en-US version of Firefox from <https://ftp.mozilla.org/pub/firefox/releases/>
and unzip and run with:

```bash
cd /opt
sudo tar xjf ~/Downloads/firefox-*.tar.bz2
/opt/firefox/firefox
```

Tested under Ubuntu 18.04 Bionic (newer Ubuntu versions have problems running
older Firefox versions, probably because of updated libraries).

Testing with `ibus-braille.py` attached to #2054, which outputs the
results of `get_surrounding_text` at each keystroke (keydown and keyup).

When things work, we get the previous value on keypress (because the new
character hasn't been output yet), and the new value on keyup.

Tested with ibus 1.5.17-3ubuntu5.3sil2.0 which has the recent surrounding text changes
(see [ppa](https://launchpad.net/~keymanapp/+archive/ubuntu/keyman-test/+sourcepub/13965871/+listing-archive-extra)).

## Test results

| Firefox version | address bar | text field |
|-----------------|-------------|------------|
| 63.0.1          | no (1)      | no (1)     |
| 104.0 (deb)     | works       | works      |
| 84.0            | no (1)      | no (1)     |
| 94.0            | no (1)      | no (1)     |
| 99.0            | works       | works      |
| 96.0            | no (1)      | no (1)     |
| 98.0            | no (1)      | no (1)     |
| 98.0.2          | no (1)      | no (1)     |
| 99.0b4          | works       | works      |
| 99.0b1          | works       | works      |

(1): works immediately after starting Firefox, both in address bar and text field
     until first page gets loaded

## Conclusion

The behavior was fixed between Firefox 98.0.2 and 99.0b1.

**Comment by mcdurdin** [comment]
I did a "quick" (I think I figured out their search interface...) check of [bugs fixed on Firefox for Linux in 99.0](https://bugzilla.mozilla.org/buglist.cgi?query_format=advanced&classification=Client%20Software&classification=Developer%20Infrastructure&classification=Components&classification=Server%20Software&classification=Other&target_milestone=99%20Branch&target_milestone=99%20Branch&target_milestone=99%20Branch&target_milestone=99%20Branch&target_milestone=99%20Branch&target_milestone=99%20Branch&target_milestone=99%20Branch&target_milestone=99%20Branch&target_milestone=99%20Branch&target_milestone=99%20Branch&op_sys=Linux&bug_status=UNCONFIRMED&bug_status=NEW&bug_status=ASSIGNED&bug_status=REOPENED&bug_status=RESOLVED&bug_status=VERIFIED&bug_status=CLOSED&resolution=---&resolution=FIXED&resolution=INVALID&resolution=WONTFIX&resolution=INACTIVE&resolution=DUPLICATE&resolution=WORKSFORME&resolution=INCOMPLETE&resolution=SUPPORT&resolution=EXPIRED&resolution=MOVED) and found the following bug which seems to be most likely to have been the fix:

* [[GTK] Failed to handle composition start signal which is sent immediately after setting IME focus](https://bugzilla.mozilla.org/show_bug.cgi?id=1753173)

**Comment by fujiwarat** [comment]
> Testing with a slightly modified version of [ibus-braille.py](https://github.com/ibus/ibus/files/9634832/ibus-braille.py.txt) originally attached to https://github.com/ibus/ibus/issues/2054, which writes the results of

I think the script is not good.
1. You don't have to call do_focus_in() in do_enable()
2. You should call get_surrounding_text() in do_focus_in()
3. You should connect "set-surrounding-text" signal instead of calling get_surrounding_text() in do_process_key_event().
